mast/integrations/utils.py

```diff
@@ -0,0 +1,386 @@
+    :param per_page: activities per page - default 100
+    :param page: pages - default 1
+    :return: JSON of list of activities
+    '''
+    url = f'https://www.strava.com/api/v3/athlete/activities'
+
+    header = {
+        'Authorization': f'Bearer {access_token}'
+    }
+    data = {
+        'before': before if not None else '',
+        'after': after,
+        'page': page,
+        'per_page': per_page
+    }
+
+    response = requests.request('GET', url, headers=header, data=data)
+    json_data = json.loads(response.text)
+    # strava_logger.info(json.dumps(json_data, indent=4))
+
+    return json_data
+
+
+def create_activity_from_strava_json(activity: dict, user: User, strava_activity_id: int) -> Activity:
+    """
+    Processes json from strava *DetailedActivity* and creates Activity instance
+    :param activity: json from strava *strava.com/activity/id*
+    :param user: User that uploaded the activity
+    :param strava_activity_id: strava activity id
+    :return: new Activity ready to save into database
+    """
+
+    distance = activity['distance']     # meters
+    time_in_secs = activity['moving_time']  # seconds
+    total_time = _get_time(time_in_secs)     # time(H:M:S)
+    elevation = activity['total_elevation_gain'] if not None else 0     # meters
+    activity_type = _get_activity_type(activity['type'])  # https://developers.strava.com/docs/reference/#api-models-ActivityType
+    name = activity['name'][:30] if not None else ''    # string
+    pace = _get_time(round(time_in_secs / distance * 1000))     # minutes per km
+    activity_date = datetime.strptime(activity['start_date'],'%Y-%m-%dT%H:%M:%SZ')  # datetime
+    score = _get_score(distance, elevation, pace, user, activity_type)     # int
+
+    # check activity constrains
+    if activity_type is None:    # unsupported ActivityType
+        strava_logger.info(
+            f'Activity :{activity_type} is not supported.')
+        return None
+
+    if _satisfy_distance_constrains(distance/1000, activity_type):     # Activity is not long enough to be counted
+        strava_logger.info(
+            f'Activity :{activity_type} of length {distance} meters does not satisfy distance constrains.')
+        return None
+
+    # creation of the activity
+    new_activity = Activity()
+    new_activity.datetime = activity_date
+    new_activity.distance = distance / 1000
+    new_activity.duration = total_time
+    new_activity.average_duration_per_km = pace
+    new_activity.type = activity_type
+    new_activity.user_id = user.id
+    new_activity.name = name
+    new_activity.elevation = elevation
+    new_activity.strava_id = strava_activity_id
+    new_activity.score = score
+
+    return new_activity
+
+
+def process_strava_webhook(data: dict):
+    """
+    Processes strava webhook
+    :param data: dictionary of data from webhook
+    :return: On Create: Activity; On Delete: None; On Update: None
+    """
```

Comment:
First, search valid user and sort out user not found because the following methods do call _get_user separately. I believe we should parse the whole strava webhook notification JSON here and then pass already valid data for further processing to other functions.